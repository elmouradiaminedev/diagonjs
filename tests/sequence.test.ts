import { expect, describe, it, beforeAll } from "vitest";
import Diagon, { DiagonType, SequenceTranslationOptions } from "src";

function generateAllOptions(): SequenceTranslationOptions[] {
  const allOptions: SequenceTranslationOptions[] = [];

  for (const asciiOnly of [true, false]) {
    for (const interpretBackSlashN of [true, false]) {
      allOptions.push({
        asciiOnly,
        interpretBackSlashN,
      });
    }
  }

  return allOptions;
}

let diagon: DiagonType;

beforeAll(async () => {
  diagon = await Diagon.init();
});

describe("Sequence expression translation", () => {
  generateAllOptions().forEach((options) => {
    describe(`With options ${JSON.stringify(options)}`, () => {
      it("should translate basic", () => {
        expect(
          diagon.translate.sequence(
            "Alice -> Bob: Hello Bob!\nAlice <- Bob: Hello Alice!",
            options,
          ),
        ).toMatchSnapshot();
      });

      it("should translate more actors", () => {
        expect(
          diagon.translate.sequence(
            "Renderer -> Browser: BeginNavigation()\nBrowser -> Network: URLRequest()\nBrowser <- Network: URLResponse()\nRenderer <- Browser: CommitNavigation()\nRenderer -> Browser: DidCommitNavigation()",
            options,
          ),
        ).toMatchSnapshot();
      });

      it("should translate actors order", () => {
        expect(
          diagon.translate.sequence(
            "Actor 2 -> Actor 3: message 1\nActor 1 -> Actor 2: message 2\n\nActor 1:\nActor 2:\nActor 3:",
            options,
          ),
        ).toMatchSnapshot();
      });

      it("should translate message order", () => {
        expect(
          diagon.translate.sequence(
            "2) Actor 2 -> Actor 3: message 1\n1) Actor 1 -> Actor 2: message 2\n\nActor 1:\nActor 2: 1<2\nActor 3:",
            options,
          ),
        ).toMatchSnapshot();
      });

      it("should translate message crossing", () => {
        expect(
          diagon.translate.sequence(
            "1) Renderer -> Browser: Message 1\n2) Renderer <- Browser: Message 2\n\nRenderer: 1<2\nBrowser: 2<1",
            options,
          ),
        ).toMatchSnapshot();
      });
    });
  });
});
