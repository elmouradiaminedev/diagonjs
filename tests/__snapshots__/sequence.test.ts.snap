// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":false} > should translate actors order 1`] = `
"┌───────┐ ┌───────┐ ┌───────┐
│Actor 1│ │Actor 2│ │Actor 3│
└───┬───┘ └───┬───┘ └───┬───┘
    │         │         │    
    │         │message 1│    
    │         │────────>│    
    │         │         │    
    │message 2│         │    
    │────────>│         │    
┌───┴───┐ ┌───┴───┐ ┌───┴───┐
│Actor 1│ │Actor 2│ │Actor 3│
└───────┘ └───────┘ └───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":false} > should translate basic 1`] = `
"┌─────┐       ┌───┐
│Alice│       │Bob│
└──┬──┘       └─┬─┘
   │            │  
   │ Hello Bob! │  
   │───────────>│  
   │            │  
   │Hello Alice!│  
   │<───────────│  
┌──┴──┐       ┌─┴─┐
│Alice│       │Bob│
└─────┘       └───┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":false} > should translate message crossing 1`] = `
" ┌────────┐┌───────┐
 │Renderer││Browser│
 └───┬────┘└───┬───┘
     │         │    
     │──┐      │    
     │Message 2│    
     │<────────│    
     │  │      │    
     │Message 1│    
     │  └─────>│    
 ┌───┴────┐┌───┴───┐
 │Renderer││Browser│
 └────────┘└───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":false} > should translate message order 1`] = `
"┌───────┐ ┌───────┐ ┌───────┐
│Actor 1│ │Actor 2│ │Actor 3│
└───┬───┘ └───┬───┘ └───┬───┘
    │         │         │    
    │message 2│         │    
    │────────>│         │    
    │         │         │    
    │         │message 1│    
    │         │────────>│    
┌───┴───┐ ┌───┴───┐ ┌───┴───┐
│Actor 1│ │Actor 2│ │Actor 3│
└───────┘ └───────┘ └───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":false} > should translate more actors 1`] = `
" ┌────────┐            ┌───────┐     ┌───────┐
 │Renderer│            │Browser│     │Network│
 └───┬────┘            └───┬───┘     └───┬───┘
     │                     │             │    
     │  BeginNavigation()  │             │    
     │────────────────────>│             │    
     │                     │             │    
     │                     │URLRequest() │    
     │                     │────────────>│    
     │                     │             │    
     │                     │URLResponse()│    
     │                     │<────────────│    
     │                     │             │    
     │ CommitNavigation()  │             │    
     │<────────────────────│             │    
     │                     │             │    
     │DidCommitNavigation()│             │    
     │────────────────────>│             │    
 ┌───┴────┐            ┌───┴───┐     ┌───┴───┐
 │Renderer│            │Browser│     │Network│
 └────────┘            └───────┘     └───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":true} > should translate actors order 1`] = `
"┌───────┐ ┌───────┐ ┌───────┐
│Actor 1│ │Actor 2│ │Actor 3│
└───┬───┘ └───┬───┘ └───┬───┘
    │         │         │    
    │         │message 1│    
    │         │────────>│    
    │         │         │    
    │message 2│         │    
    │────────>│         │    
┌───┴───┐ ┌───┴───┐ ┌───┴───┐
│Actor 1│ │Actor 2│ │Actor 3│
└───────┘ └───────┘ └───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":true} > should translate basic 1`] = `
"┌─────┐       ┌───┐
│Alice│       │Bob│
└──┬──┘       └─┬─┘
   │            │  
   │ Hello Bob! │  
   │───────────>│  
   │            │  
   │Hello Alice!│  
   │<───────────│  
┌──┴──┐       ┌─┴─┐
│Alice│       │Bob│
└─────┘       └───┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":true} > should translate message crossing 1`] = `
" ┌────────┐┌───────┐
 │Renderer││Browser│
 └───┬────┘└───┬───┘
     │         │    
     │──┐      │    
     │Message 2│    
     │<────────│    
     │  │      │    
     │Message 1│    
     │  └─────>│    
 ┌───┴────┐┌───┴───┐
 │Renderer││Browser│
 └────────┘└───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":true} > should translate message order 1`] = `
"┌───────┐ ┌───────┐ ┌───────┐
│Actor 1│ │Actor 2│ │Actor 3│
└───┬───┘ └───┬───┘ └───┬───┘
    │         │         │    
    │message 2│         │    
    │────────>│         │    
    │         │         │    
    │         │message 1│    
    │         │────────>│    
┌───┴───┐ ┌───┴───┐ ┌───┴───┐
│Actor 1│ │Actor 2│ │Actor 3│
└───────┘ └───────┘ └───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":false,"interpretBackSlashN":true} > should translate more actors 1`] = `
" ┌────────┐            ┌───────┐     ┌───────┐
 │Renderer│            │Browser│     │Network│
 └───┬────┘            └───┬───┘     └───┬───┘
     │                     │             │    
     │  BeginNavigation()  │             │    
     │────────────────────>│             │    
     │                     │             │    
     │                     │URLRequest() │    
     │                     │────────────>│    
     │                     │             │    
     │                     │URLResponse()│    
     │                     │<────────────│    
     │                     │             │    
     │ CommitNavigation()  │             │    
     │<────────────────────│             │    
     │                     │             │    
     │DidCommitNavigation()│             │    
     │────────────────────>│             │    
 ┌───┴────┐            ┌───┴───┐     ┌───┴───┐
 │Renderer│            │Browser│     │Network│
 └────────┘            └───────┘     └───────┘
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":false} > should translate actors order 1`] = `
".-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
    |         |         |    
    |         |message 1|    
    |         |-------->|    
    |         |         |    
    |message 2|         |    
    |-------->|         |    
.-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":false} > should translate basic 1`] = `
".-----.       .---.
|Alice|       |Bob|
'-----'       '---'
   |            |  
   | Hello Bob! |  
   |----------->|  
   |            |  
   |Hello Alice!|  
   |<-----------|  
.-----.       .---.
|Alice|       |Bob|
'-----'       '---'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":false} > should translate message crossing 1`] = `
" .--------..-------.
 |Renderer||Browser|
 '--------''-------'
     |         |    
     |--.      |    
     |Message 2|    
     |<--------|    
     |  |      |    
     |Message 1|    
     |  '----->|    
 .--------..-------.
 |Renderer||Browser|
 '--------''-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":false} > should translate message order 1`] = `
".-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
    |         |         |    
    |message 2|         |    
    |-------->|         |    
    |         |         |    
    |         |message 1|    
    |         |-------->|    
.-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":false} > should translate more actors 1`] = `
" .--------.            .-------.     .-------.
 |Renderer|            |Browser|     |Network|
 '--------'            '-------'     '-------'
     |                     |             |    
     |  BeginNavigation()  |             |    
     |-------------------->|             |    
     |                     |             |    
     |                     |URLRequest() |    
     |                     |------------>|    
     |                     |             |    
     |                     |URLResponse()|    
     |                     |<------------|    
     |                     |             |    
     | CommitNavigation()  |             |    
     |<--------------------|             |    
     |                     |             |    
     |DidCommitNavigation()|             |    
     |-------------------->|             |    
 .--------.            .-------.     .-------.
 |Renderer|            |Browser|     |Network|
 '--------'            '-------'     '-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":true} > should translate actors order 1`] = `
".-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
    |         |         |    
    |         |message 1|    
    |         |-------->|    
    |         |         |    
    |message 2|         |    
    |-------->|         |    
.-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":true} > should translate basic 1`] = `
".-----.       .---.
|Alice|       |Bob|
'-----'       '---'
   |            |  
   | Hello Bob! |  
   |----------->|  
   |            |  
   |Hello Alice!|  
   |<-----------|  
.-----.       .---.
|Alice|       |Bob|
'-----'       '---'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":true} > should translate message crossing 1`] = `
" .--------..-------.
 |Renderer||Browser|
 '--------''-------'
     |         |    
     |--.      |    
     |Message 2|    
     |<--------|    
     |  |      |    
     |Message 1|    
     |  '----->|    
 .--------..-------.
 |Renderer||Browser|
 '--------''-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":true} > should translate message order 1`] = `
".-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
    |         |         |    
    |message 2|         |    
    |-------->|         |    
    |         |         |    
    |         |message 1|    
    |         |-------->|    
.-------. .-------. .-------.
|Actor 1| |Actor 2| |Actor 3|
'-------' '-------' '-------'
"
`;

exports[`Sequence expression translation > With options {"asciiOnly":true,"interpretBackSlashN":true} > should translate more actors 1`] = `
" .--------.            .-------.     .-------.
 |Renderer|            |Browser|     |Network|
 '--------'            '-------'     '-------'
     |                     |             |    
     |  BeginNavigation()  |             |    
     |-------------------->|             |    
     |                     |             |    
     |                     |URLRequest() |    
     |                     |------------>|    
     |                     |             |    
     |                     |URLResponse()|    
     |                     |<------------|    
     |                     |             |    
     | CommitNavigation()  |             |    
     |<--------------------|             |    
     |                     |             |    
     |DidCommitNavigation()|             |    
     |-------------------->|             |    
 .--------.            .-------.     .-------.
 |Renderer|            |Browser|     |Network|
 '--------'            '-------'     '-------'
"
`;
